mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td

```diff
@@ -204,6 +204,15 @@ def ROCDL_ReadlaneOp : ROCDL_IntrOp<"readlane", [], [0], [AllTypesMatch<["res",
    }];
 }
 
+// The LLVM intrinsic function name is rather mouthful,
+// so here we opt to use a shorter rocdl name.
+def ROCDL_WaveIdOp :  LLVM_IntrOpBase<ROCDL_Dialect, "wave_id",
+                        "amdgcn_s_get_waveid_in_workgroup", [], [], [], 1>,
```

Comment:
We should still be able to mark this as speculatable and with read memory effects only

Reply:
Actually, the corresponding llvm intrinsic function has properties as follows:
```
[NoUndef<RetIndex>, IntrNoMem, IntrHasSideEffects, IntrWillReturn, IntrNoCallback, IntrNoFree]>;
```

and also:
```
def Pure : TraitList<[AlwaysSpeculatable, NoMemoryEffect]>;
```

So I guess we should keep `Pure` ?

Reply:
If it was pure, you would be able to hoist it out of parallel loops before distribution to subgroups. I suspect it may be `IntrNoMem` in llvm because they can't represent higher-level programs

Reply:
I guess we cannot make it speculatable because it has side effects.

Reply:
?

Reply:
@kuhar if we make the instruction speculatable then it should not have side effects, but the LLVM intrinsic has `IntrHasSideEffects`, and `IntrNoMem`, so I think the only applicable trait here is `NoMemoryEffect`.

Reply:
See https://mlir.llvm.org/docs/Rationale/SideEffectsAndSpeculation/

Reply:
@kuhar so it is only conditionally speculatable in that it cannot be hoisted out of parallel loops but speculatable within parallel loops. Something like this might help:

```C++
    ::mlir::Speculation::Speculatability
    WaveIdOp::getSpeculatability() {
      Operation* enclosingParallelLoopOp = 
          getParentOfType<::mlir::scf::ForallOp>();
      return enclosingParallelLoopOp ?
          ::mlir::Speculation::Speculatable : ::mlir::Speculation::NotSpeculatable;
    }
```

But question: should this go into ROCDL op definitions? I feel it will be a bit out of place, as the restriction is downstream-specific.

Any thoughts on this?

Reply:
I'd make it always speculatable but specify read memory effects. Can we do that without specifying what memory is being read? Otherwise just leave it without any annotation IMO.

Reply:
One thing I'd want to get a sense of is _why_ the LLVM-side intrinsic is marked as having side effects - I suspect it's for internal reasons about needing to fetch dimensions from some memory, but I'd want to look into that.

Conceptually, "wave ID in subgroup" should have the same purity as `gpu.thread_id`

Reply:
`gpu.thread_id` and `gpu.subgroup_id` are both pure.

Reply:
Also, as a side note, the "major version >= 10" check is, going off of `FeatureGetWaveIdInst` down in LLVM, wrong. This instruction only seems to appear on gfx101x, after which time it was removed.

I think the lowering for `subgroup_id` should be ID in workgroup % subgroup size

Reply:
>  This instruction only seems to appear on gfx101x, after which time it was removed.

Huh, I thought the instruction would stay in future architectures.

Reply:
We can expand to `thread_id / wave_sz`, no?

Reply:
since `subgroup_id` will be expanded differently so I removed this definition of op.